### PR TITLE
Follow-up 1726. Update optuna sweeper dependency and some cleanups.

### PR DESCRIPTION
**Pull request by toshihikoyanase**
## Motivation

This PR is a follow-up for #1726. Since #1726 has not merged yet, this PR includes the changes of the original PR.

*Changes*
- Set lower bound of Optuna version (>=2.5.0). Since the code in #1726 uses `study.ask` and `study.tell`, it does not work with `optuna<2.5.0`. See https://github.com/facebookresearch/hydra/pull/1726#discussion_r679915412.
- Update the scenario of `test_optuna_example` to save computation time. It also makes the test robust to Optuna's changes. See  "Test Plan" for more details.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/facebookresearch/hydra/blob/master/CONTRIBUTING.md)?

Yes

## Test Plan

This PR changes the test scenario of `test_optuna_example`. The current test function tries to find the global optima by the random sampler, but it requires many trials (about 1000 trials) to make it robust to the change of seed values or the sampler
implementation.

The new test function compares the results in Optuna's storage and the results returned by the plugin. I think this is a straight forward way to verify the plugin behavior. Since it does not require the global optima, we can use the same number of trials as the current master.

(See also https://github.com/facebookresearch/hydra/pull/1726#discussion_r678299458)

## Related Issues and PRs

The original PR is #1726 .

### CODE REVIEW

File: plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py
Comment (omry) [inline plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py:132]:
```suggestion
        f"hydra.sweeper.storage={storage}",
        f"hydra.sweeper.study_name={study_name}",
```

File: plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py
Comment (omry) [inline plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py:152]: Can we use constants here (with a reasonable epsilon?)
The current testing looks like it will be too lenient to  functional issues in Optuna.

File: plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py
Comment (toshihikoyanase) [reply plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py]: Thanks!

File: plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py
Comment (toshihikoyanase) [reply plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py:152]: Thank you for your comment.

I understand your concern since the current Optuna repository does not have CI jobs which execute search performance tests.

I tried to estimated the reasonable epsilon with [a notebook](https://colab.research.google.com/drive/1HIvjys_NhinskSR1AFlby0NsLstRlZUD?usp=sharing). I repeated 1000 time of random search (that is the search algorithm used in this example), and found that the 95th percentile value of the best values was 6.25.  

The acceptance range is shown in the following contour plot with the red rectangle. I guess it is reasonably small considering that the search algorithm was random search. What do you think?

image

The suggested change is as follows:
```python
    # 95th percentile with 1000 different seed values.
    assert returns["best_value"] <= 6.25
```

Comment (omry) [reply plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py:152]: Thanks for evaluating 
Seems like a very large range (and also, if I understand correctly, 95th percentile means it will fail with one in 20 random seeds).

Is there an algorithm that converges faster than a random search we can test with?

Comment (toshihikoyanase) [reply plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py:152]: I applied the following three changes to the test case.
1. Employ 99the percentile (it will fail with one in 100 random seeds.)
2. Use TPE sampler instead of the random search
3. Increase the number of trials from 20 to 40.

The threshold values were changed as follows:
1. `6.25` to `12.25`
2. `12.25` to `2.27`
3. `2.27` to `0.25`

So, I applied all changes in the commit https://github.com/facebookresearch/hydra/pull/1746/commits/268cbb7ae399d611c5e88e38b0dadbfb200e7886.

The change increased the computation time, but it was still shorter than that of random search with 100 trials.
```
random with 20 trials
2.90s call     plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py::test_optuna_example[False]
2.85s call     plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py::test_optuna_example[True]

random with 100 trials (from #1726)
8.89s call     plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py::test_optuna_example[False]
8.79s call     plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py::test_optuna_example[True]

tpe with 20 trials
2.97s call     plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py::test_optuna_example[False]
2.95s call     plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py::test_optuna_example[True]

tpe with 40 trials
5.15s call     plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py::test_optuna_example[False]
4.82s call     plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py::test_optuna_example[True]
```

Comment (omry) [reply plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py:152]: Given that the seed is fixed in the test, I think 95th percentile is acceptable.
What bounds do we get for 95th percentile, tpe and 20 iterations?

Go ahead and proceed with whatever values you are comfortable with after this evaluation. I think it will be a good idea to add a comment to this discussion to the test.

Comment (toshihikoyanase) [reply plugins/hydra_optuna_sweeper/tests/test_optuna_sweeper_plugin.py:152]: Thank you for your suggestion.
The 95th percentile with TPE and 20 iterations is 2.27.
I updated the bound and added a code comment on this discussion in this commit https://github.com/facebookresearch/hydra/pull/1746/commits/e463885adacf2713ddee97f3ef8d381633ff57ed

**Review by omry — APPROVED** [review]
lgtm. thanks.

Jieru, can you take over from here?

**Comment by jieru-hu** [comment]
so sorry for getting back to this late! the PR lgtm. Thanks so much!
 @toshihikoyanase  could you rebase master again? I will merge this after that. 

Thanks again!

**Comment by toshihikoyanase** [comment]
@jieru-hu Thank you for your review! I rebased master, but the CI failed at the test of nevergrad as follows:

```console
E   nevergrad.common.errors.InefficientSettingsWarning: DE algorithms are inefficient with budget < 60
=========================== short test summary info ===========================
ERROR tests\test_nevergrad_sweeper_plugin.py - nevergrad.common.errors.Ineffi...
!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!
============================== 1 error in 1.74s ===============================
```
Do you have any ideas about it?

**Comment by jieru-hu** [comment]
Thanks @toshihikoyanase 
Looks like the tests were broken by a new warnings introduced by the new nevergrad version released yesterday. 
I will fix the tests and let you know.

**Comment by jieru-hu** [comment]
@toshihikoyanase  the CI has been fixed on the main branch. Could you try rebase? Sorry for the hassle!

**Comment by toshihikoyanase** [comment]
@jieru-hu Thank you for resolving the CI issue! All the CI jobs have passed. Please take a look.

**Comment by jieru-hu** [comment]
many thanks @toshihikoyanase !!

**Pull request merged**
